### ISSUE TITLE
Feature Request: Postgres/Ecto support

### ISSUE THREAD

**Issue by asummers** [tompave/fun_with_flags#1]
Currently Molasses is the only feature flag library in Elixir available that has Postgres backing as an option, though I consider the abstraction of groups much cleaner in this library. Would it be possible  to add a persistence layer for Postgres? 

Looking through the code, I was looking for where that would get hooked in. as some sort of generic "Store Protocol/Behaviour", to see if that would even be possible with the current implementation, but the current stores (simple_store.ex and store.ex) don't seem to reference one. `FunWithFlags.Store.Persistent` seems to be hard coded to start a Redis worker, so perhaps that being made into a pluggable interface of sorts would more easily allow this, potentially opening the door to Mongo adapters or RethinkDB adapters or even Application config backed adapters.

**Comment by tompave** [comment]
Hi, and thank you for the suggestion and for liking the API :-).

You are correct, `FunWithFlags.Store.Persistent` is meant to be the storage module. I should probably either rename it with something Redis-specific or extract the Redis bit and make it pluggable, with a standard interface. At the moment that module is still private and undocumented, so it would be doable.

If that was done, adding support for different persistent datastores would be trivial, as the only requirement is to be able to store and retrieve string key-value pairs.

What would be more difficult to implement is the pub-sub cache busting, which at the moment relies on Redis pub-sub [for the reasons discussed here](https://github.com/tompave/fun_with_flags#why-not-distributed-erlang). So using a Postgres persistence adapter would either require to disable the ETS cache or to also add an adapter for the [`Notifications`](https://github.com/tompave/fun_with_flags/blob/master/lib/fun_with_flags/notifications.ex#L49) event listener module. This notifications adapter should use Postgres' own [pub-sub system](https://www.postgresql.org/docs/9.6/static/libpq-notify.html), built on the [`NOTIFY`](https://www.postgresql.org/docs/9.6/static/sql-notify.html) and [`LISTEN`](https://www.postgresql.org/docs/9.6/static/sql-listen.html) commands.

Of the two approaches I would recommend the latter, because I don't think that storing feature toggles in a DB is a good idea. In my experience, this kind of DB adapters make sense for small and low-traffic applications, or to get started quickly with new projects when one doesn't want to add another external dependency (Redis). I appreciate that it's a legitimate use case, however.

Let me explain my previous point a bit better, to help you and other decide if using Postgres instead of Redis is the right choice.

While disabling the ETS cache might be fine for some applications, it would of course increase the load on the persistent data store. This would be fine with Redis (and actually _Redis itself_ is usually used as a cache), but I wouldn't recommend it with a relational DB.

Once they are added to a project, it's easy to start using feature flags more and more (which is a good thing!), and they will need to be read very frequently. In a web application, loading more than one flag per HTTP request is common.
This is the kind of access pattern that people actively try to cache to not overload the DB, typically using Redis or Memcached. Storing the flags in the DB without a cache layer introduces the risk to quickly erode the DB resources that should be used for "real application data".

**Comment by tompave** [comment]
[Here](https://github.com/tompave/fun_with_flags#origin) you can find another explanation of why this library uses an ETS cache.

**Comment by asummers** [comment]
In my ideal world, I would have the cache periodically refresh from the database (some sort of GET on the flag table), and load it into ETS. That way you're only paying the penalty every `interval`, say 5 minutes or whatever is appropriate. Using Listen/Notify would be a good improvement to that, but shouldn't be strictly necessary. Using a timed refresh could lead to distributed servers having different flags, but the system should be resilient to that from the application level. 

Within our application, we have a larger eventing system using Phoenix Pub Sub that we would be able to hook into to know that the flag values were updated, which would eliminate the need for the notify listen. On system start up, it would ping the database, load configs, then have another process somewhere ready to update ETS when the time comes. In general, I agree with you -- hitting the DB every time you wanna check a flag is very heavy -- but that won't be the reality for us even without the listen/notify being added here.

**Comment by asummers** [comment]
I guess all I'm really saying in the above is that the persistent storage and the cache could be thought of completely independently, and might not be up to this library in every case. Being able to override how the caching mechanism works would be powerful.

**Comment by tompave** [comment]
> In my ideal world, I would have the cache periodically refresh from the database (some sort of GET on the flag table), and load it into ETS. That way you're only paying the penalty every interval, say 5 minutes or whatever is appropriate.
> (...)
> On system start up, it would ping the database, load configs, then have another process somewhere ready to update ETS when the time comes.

I see. That's actually already builtin in `FunWithFlags.Store.Cache`, where [a TTL is used](https://github.com/tompave/fun_with_flags/blob/master/lib/fun_with_flags/store/cache.ex#L21-L39) (you can [configure it](https://github.com/tompave/fun_with_flags#configuration)).
When caching is enabled, looking up a flag will check the ETS cache first, and if the value is missing or expired the flag will be looked up in the persistent Redis store. This is mainly a fail-safe in case the pub-sub cache busting doesn't work for some reason (e.g. the listener process crashes and the pub-sub notification is published while it's restarting).
At the moment flags are loaded lazily the first time they’re requested, and again when they’re checked after they’ve expired, which helps evenly distribute the Redis lookups over time.


> Using Listen/Notify would be a good improvement to that, but shouldn't be strictly necessary. Using a timed refresh could lead to distributed servers having different flags, but the system should be resilient to that from the application level.

You are describing a very relaxed eventual consistency model. I agree that it makes sense for some applications, but a design goal of this library is to try to be as consistent as possible within the limitations of distributed systems over a network.
So while it’s true that the pub-sub cache-busting doesn’t really guarantee that the state is immediately consistent across nodes, it still allows the nodes to have “low-latency” eventual consistency.

I think that for feature flags this is important (and it’s the main reason why I included the option to disable the cache), and from what you say I think that your requirements are a bit different from how I used feature toggles in the past.
For example, let’s say that you enable a feature flag for a new UI header design. You _need_ it to be consistent across servers and nodes. Otherwise, as your users refresh the page, and the HTTP requests are served by different nodes, they’ll see different and conflicting UIs with each page load. The same applies to backend logic, where consistency could be even more critical. Think about a feature toggle that changes the price for a service or store item: you don’t want this to change randomly while your users are going through a checkout flow.

**Comment by tompave** [comment]
> Within our application, we have a larger eventing system using Phoenix Pub Sub that we would be able to hook into to know that the flag values were updated, which would eliminate the need for the notify listen.
> (...)
> I guess all I'm really saying in the above is that the persistent storage and the cache could be thought of completely independently, and might not be up to this library in every case. Being able to override how the caching mechanism works would be powerful.

Yes, using Phoenix’s PubSub system (with PG2 or Redis backends) is a very valid alternative.
The main reason I didn’t use it in this library is that I didn’t want to add the dependency, as FunWithFlags could be used without Phoenix.


Your suggestions are interesting, however.
I think I should definitely refactor `FunWithFlags.Store.Persistent` to start, and maybe `FunWithFlags.Notifications`, so that Redis can be swapped with something else.
This should allow you to replace Redis with Ecto, and you can write your version of [the notification listener logic](https://github.com/tompave/fun_with_flags/blob/master/lib/fun_with_flags/notifications.ex#L67-L95) to work with your internal pub-sub system.


As for this project, I don’t think that I’m going to add adapters other than Redis, at least for the moment. The reason is that there are other features on the roadmap that I want to complete before considering the internal API and storage format stable:

* a web GUI to toggle the flags, as a Plug (standalone and/or pluggable in Phoenix);
* %-based gates.


Before that, however, I think I’ll do the refactoring I mentioned above, so that it will be possible to develop 3rd party extensions like your Postgres adapter.

Thanks for the suggestion!

**Comment by tompave** [comment]
> If I complete that, it would be possible to develop 3rd party extensions like your Postgres adapter, so I think I should do it at some point.

Ok, so it sounded useful enough and it didn't require a lot of changes.
It's merged into master, I'll release it with the next version.

The internal API for persistence adapters is:
```elixir
```

The configuration API is:
```elixir
config :fun_with_flags, :persistence_adapter, FunWithFlags.Store.Persistent.MySpecialAdapter
```

**Comment by tompave** [comment]
Closing as done.

**Comment by tompave** [comment]
This is now available in version `0.7.0`, on Hex: https://hex.pm/packages/fun_with_flags/0.7.0

**Comment by tompave** [comment]
Hey @asummers, a new Ecto adapter has just been released in version `0.9.0`.
